### PR TITLE
[SPARK-28033][SQL] String concatenation should lower precedence than other operators

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

This PR makes string concatenation(`'||'`) lower precedence than the other operators for being consistent with other databases, for example:
```sql
spark-sql> select 'four: ' || 2 + 2;
NULL
spark-sql> select 'zero: ' || 2 ^ 2;
Error in query: cannot resolve '(CAST(concat('zero: ', CAST(2 AS STRING)) AS DOUBLE) ^ 2)' due to data type mismatch: differing types in '(CAST(concat('zero: ', CAST(2 AS STRING)) AS DOUBLE) ^ 2)' (double and int).; line 1 pos 7;
'Project [unresolvedalias((cast(concat(zero: , cast(2 as string)) as double) ^ 2), None)]
+- OneRowRelation
```

PostgreSQL, Vertica, Hive, Teradata support this feature.
Oracle's string concatenation has same precedence with [`+` and `-`](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/About-SQL-Operators.html#GUID-FEF44762-F45C-41D9-B380-F6A61AD25338).
Presto, SQL Server, DB2 and MySQL does not support it.
**PostgreSQL**:
```sql
postgres=# select substr(version(), 0, 16), 'four: ' || 2 + 2, 'zero: ' || 2 ^ 2;
     substr      | ?column? | ?column?
-----------------+----------+----------
 PostgreSQL 11.3 | four: 4  | zero: 4
(1 row)
```
**Vertica**:
```sql
dbadmin=> select version(), 'four: ' || 2 + 2, 'zero: ' || 2 ^ 2;
              version               | ?column? | ?column?
------------------------------------+----------+----------
 Vertica Analytic Database v9.1.1-0 | four: 4  | zero: 4
(1 row)
```
**Hive**:
```sql
hive> select version(), 'four: ' || 2 + 2, 'zero: ' || 2 ^ 2;
OK
3.1.1 rf4e0529634b6231a0072295da48af466cf2f10b7	four: 4	zero: 0
Time taken: 14.368 seconds, Fetched: 1 row(s)
```
**Teradata**:
image
**Oracle**:
image

**Presto** throws `Unexpected parameters (varchar(6), integer) for function concat`.
**SQL Server** throws `Incorrect syntax near '|'`
**DB2** throws `Invalid character found in a character string argument of the function "DECFLOAT"`
**MySQL** returns unexpected result:
```sql
mysql> select version(), 'four: ' || 2 + 2, 'zero: ' || 2 ^ 2;
+-----------+-------------------+-------------------+
| version() | 'four: ' || 2 + 2 | 'zero: ' || 2 ^ 2 |
+-----------+-------------------+-------------------+
| 5.7.26    |                 1 |                 0 |
+-----------+-------------------+-------------------+
```

## How was this patch tested?

unit tests

### CODE REVIEW

**Comment by mgaido91** [comment]
I think this makes sense, but since this is a breaking change, shall we add a note in the migration guide?

**Review by dongjoon-hyun — COMMENTED** [review]
+1 for @mgaido91 's suggestion.
Also, cc @maropu since this is added at b526f70c16caa2062a77117808860e1d3662114c, `[SPARK-19951][SQL] Add string concatenate operator || to Spark SQL`.
Also, cc @gatorsmile since this is a part of PostgreSQL compatibility, [SPARK-27764](https://issues.apache.org/jira/browse/SPARK-27764).

BTW, @wangyum . Could you add more database results like `Presto`, please?

**Comment by wangyum** [comment]
Thank you @dongjoon-hyun I add Vertica, Teradata, Oracle, Presto, SQL Server, DB2 and MySQL.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @wangyum .

**Comment by wangyum** [comment]
@mgaido91 @dongjoon-hyun  Should we update the migration guide? I think this is a bug fix, not a behavior change.

**Comment by mgaido91** [comment]
I do think so, it is a behavior change. Despite the previous behavior was wrong, this may change users' results so we should notify them about the change, in order to let them check their queries using this operator. It would be pretty bad for a user to realize that the query has a behavior change and there is no documentation at all about what changed...

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -668,10 +668,11 @@ valueExpression
     : primaryExpression                                                                      #valueExpressionDefault
```
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Could you add a comment above line 667? 
> Note: check the operator precedence when adding the new operator.

**Comment by gatorsmile** [comment]
I think we are following Oracle when adding the concatenation.
https://docs.oracle.com/cd/E11882_01/server.112/e41084/operators001.htm#SQLRF51153

Could you check whether it is true in Oracle?

**Comment by gatorsmile** [comment]
We need to document our own operator precedence. Welcome the contribution!

**Comment by wangyum** [comment]
It seems that it works only when the other side is a number string or a space:
```sql
SQL> select banner from v$version;

BANNER
--------------------------------------------------------------------------------
Oracle Database 12c Standard Edition Release 12.2.0.1.0 - 64bit Production
PL/SQL Release 12.2.0.1.0 - Production
CORE	12.2.0.1.0	Production
TNS for Linux: Version 12.2.0.1.0 - Production
NLSRTL Version 12.2.0.1.0 - Production

SQL> select '4' || 2 + 2 from dual;

  '4'||2+2
----------
	44

SQL> select '44' || 2 + 2 from dual;

 '44'||2+2
----------
       444

SQL> select '' || 2 + 2 from dual;

   ''||2+2
----------
	 4

SQL> select '4a' || 2 + 2 from dual;
select '4a' || 2 + 2 from dual
             *
ERROR at line 1:
ORA-01722: invalid number

SQL> select ':' || 2 + 2 from dual;
select ':' || 2 + 2 from dual
            *
ERROR at line 1:
ORA-01722: invalid number

SQL> select 'four: ' || 2 + 2 from dual;
select 'four: ' || 2 + 2 from dual
                 *
ERROR at line 1:
ORA-01722: invalid number
```

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: OK.

**Comment by gatorsmile** [comment]
It is hard to say which precedence orders [Oracle or PostgreSQL] are better. Your example only shows our auto type casting has inconsistent behaviors.

File: docs/sql-migration-guide-upgrade.md [hunk 2]
```diff
@@ -23,6 +23,8 @@ license: |
 {:toc}
 
 ## Upgrading From Spark SQL 2.4 to 3.0
+  - Since Spark 3.0, the string concatenation operator(`||`) has lower precedence than other operators.
```
Comment (maropu) [inline docs/sql-migration-guide-upgrade.md]: Is it not "has the lowest precedence among operators"? Also, could you add a simple example query in this statement?

**Pull request closed without merging**